src/PrestaShopBundle/Controller/Api/TranslationController.php

```diff
@@ -326,90 +358,67 @@ protected function guardAgainstInvalidTranslationResetRequest($content)
-
-        $moduleProvider = $this->container->get('prestashop.translation.external_module_provider');
-        $moduleProvider->setModuleName($selectedModuleName);
-
-        $treeBuilder = new TreeBuilder($locale, $theme);
-        $catalogue = $treeBuilder->makeTranslationArray($moduleProvider, $search);
-
-        return $this->getCleanTree($treeBuilder, $catalogue, $theme, $search, $selectedModuleName);
-    }
-
-    /**
-     * @param string $lang Two-letter iso code
-     * @param null $search
      *
-     * @return array
+     * @throws Exception
      */
-    private function getMailsSubjectTree($lang, $search = null)
+    private function getTree(string $lang, string $type, array $search, ?string $theme = null, ?string $module = null)
     {
-        $theme = null;
+        $locale = $this->translationService->langToLocale($lang);
 
-        $treeBuilder = new TreeBuilder($this->translationService->langToLocale($lang), $theme);
-        $catalogue = $this->translationService->getTranslationsCatalogue($lang, 'mails', $theme, $search);
+        $catalogue = $this->translationService->getTranslationsCatalogue(
+            $this->buildProviderType($type, $theme, $module),
+            $locale,
+            $search
+        );
 
-        return $this->getCleanTree($treeBuilder, $catalogue, $theme, $search);
-    }
+        $apiBuilder = new TranslationApiTreeBuilder($this->container->get('router'), new TreeBuilder());
 
-    /**
-     * @param string $lang Two-letter iso code
-     * @param null $search
-     *
-     * @return array
-     */
-    private function getMailsBodyTree($lang, $search = null)
-    {
-        $theme = null;
-
-        $treeBuilder = new TreeBuilder($this->translationService->langToLocale($lang), $theme);
-        $catalogue = $this->translationService->getTranslationsCatalogue($lang, 'mails_body', $theme, $search);
-
-        return $this->getCleanTree($treeBuilder, $catalogue, $theme, $search);
+        return $apiBuilder->buildDomainTreeForApi(
+            $catalogue,
+            $locale,
+            $theme,
+            $search
+        );
     }
 
     /**
-     * Make final tree.
-     *
-     * @param TreeBuilder $treeBuilder
-     * @param $catalogue
+     * @param string $type
      * @param string|null $theme
-     * @param string|null $search
      * @param string|null $module
      *
-     * @return array
+     * @return TypeInterface
      */
-    private function getCleanTree(TreeBuilder $treeBuilder, $catalogue, $theme, $search = null, $module = null)
-    {
-        $translationsTree = $treeBuilder->makeTranslationsTree($catalogue);
-        $translationsTree = $treeBuilder->cleanTreeToApi($translationsTree, $this->container->get('router'), $theme, $search, $module);
-
-        return $translationsTree;
+    private function buildProviderType(
+        string $type,
+        ?string $theme = null,
+        ?string $module = null
+    ): TypeInterface {
```

Comment:
```suggestion
    private function buildProviderType(
        string $type,
        ?string $name = null
    ): TypeInterface {
```

Could even be `$config = ['name' => '']`. regarding the method content, the $module and $theme can't have a value as the same time.

Reply:
up

Reply:
Did the modification to have a single parameter